src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
+
+class AstPredicate {
+  public:
+    AstPredicate(
+        std::string name,
+        std::vector<std::string> args,
+        AstSign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        sign_(std::move(sign)) {}
+  private:
+    std::string name_;
+    std::vector<std::string> args_;
+    AstSign sign_;
+};
+
+// This represents an expression that applies either a predicate or a 
+// "canActAs" to a principal, and it is a sub-term of AstFlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class AstVerbPhrase {
+  public:
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPredicate pred);
+    std::unique_ptr<AstVerbPhrase>
+      CreateVerbPhrase(AstPrincipal right_prin);
```

Comment:
Factory methods that create an object must be declared `static`. Otherwise, you need to call them on a particular instance, but you don't yet have an instance upon which to call them.

Reply:
Do we need the factory methods or can we just have constructors as mentioned below?

Also, why do you need to return a `unique_ptr` and not an object itself?

Reply:
I used them so that these are all created with smart pointers, and so that we always get the memory management features from smart pointers (prevention of leaks and double frees). It also looked like this pattern was used in a few cases in the other IR code for Raksha.

For now I kept the factory methods and made them `static`. Let me know if you'd rather I get rid of these and just make the constructors public. I have no opinion about this.

Reply:
In this case, defining public constructors would be more versatile. We can have both value types as well as unique_ptr if needed. See my comment related to `VerbPhrase` definition.